{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "083d746d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gosling as gos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6177818e",
   "metadata": {},
   "source": [
    "## Convert GFF3 to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ef5ebd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gffpandas.gffpandas as gffpd\n",
    "import pandas as pd\n",
    "\n",
    "gff_filename = 'MZ781228.gff3' \n",
    "\n",
    "annotation = gffpd.read_gff3(gff_filename)\n",
    "attr_to_columns = annotation.attributes_to_columns()\n",
    "\n",
    "# attr_to_columns['hs'] = 'hs' + (attr_to_columns.index + 1).astype(str)\n",
    "attr_to_columns['gene_prefix'] = attr_to_columns.apply(lambda row: row['gbkey'] if pd.isna(row['gene']) \n",
    "                            else (row['gene'][:3] if pd.notna(row['gene']) else ''), axis=1)\n",
    "\n",
    "attr_to_columns.to_csv(f'{gff_filename}.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd39bd85",
   "metadata": {},
   "source": [
    "## Import Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "72c6202d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = gos.csv(\n",
    "  url=\"./MZ781228.gff3.csv\",\n",
    "  chromosomeField=\"seq_id\",\n",
    "  chromosomePrefix=\"MZ781228.\",\n",
    "  genomicFields=[\"start\", \"end\"]\n",
    ")\n",
    "\n",
    "data_base = gos.csv(\n",
    "    url=\"./MZ781228.fasta.csv\"\n",
    ")\n",
    "\n",
    "data_gc = gos.csv(\n",
    "    url=\"./MZ781228_gc_content.csv\",\n",
    ")\n",
    "\n",
    "data_gc_skew = gos.csv(\n",
    "    url=\"./MZ781228_gc_skew.csv\",\n",
    ")\n",
    "\n",
    "data_coverage = gos.csv(\n",
    "    url=\"./MZ781228_sort_out.12F.bam.tsv\",\n",
    "    separator=\"\\t\",\n",
    "    headerNames=[\"id\", \"position\", \"depth\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b69c7365",
   "metadata": {},
   "outputs": [],
   "source": [
    "track_base = gos.Track(data).encode(\n",
    "    tooltip=[\"start:G\", \"end:G\", \"strand:N\", \"gene:N\"]\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ")\n",
    "\n",
    "track_gene_base = track_base.mark_rect().encode(\n",
    "    color=gos.Color(\"gene_prefix:N\",\n",
    "            domain=[\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"],\n",
    "            range=[\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ")\n",
    "\n",
    "track_gene_posstrand = track_gene_base.transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"-\"]\n",
    ").properties(\n",
    "    title=\"Positive Strand\",\n",
    ")\n",
    "\n",
    "track_gene_negstrand = track_gene_base.transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"+\"]\n",
    ").properties(\n",
    "    title=\"Negative Strand\",\n",
    ")\n",
    "\n",
    "label_gene_base = track_base.mark_text(dy=15).encode(\n",
    "    x=gos.X(\"start:G\", axis=\"top\"),\n",
    "    xe=\"end:G\",\n",
    "    text=\"gene:N\",\n",
    "    size=gos.value(15)\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").visibility_lt(\n",
    "    measure=\"width\",\n",
    "    threshold=\"|xe-x|\",\n",
    "    transitionPadding=10,\n",
    "    target=\"mark\",\n",
    ")\n",
    "\n",
    "label_gene_posstrand = label_gene_base.transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"-\"]\n",
    ")\n",
    "\n",
    "label_gene_negstrand = label_gene_base.transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"+\"]\n",
    ")\n",
    "\n",
    "track_gc = gos.Track(data_gc).mark_area().encode(\n",
    "    # color=gos.Color(field=\"gc_content\",\n",
    "    #         type=\"quantitative\",\n",
    "    #         domain=[16.3, 14.7, 14.0],\n",
    "    #         range=[\"#8c510a\", \"#d8b365\", \"#f6e8c3\"],\n",
    "    #     ),\n",
    "    color=gos.value(\"steelBlue\"),\n",
    "    x=gos.X(\"position:G\"),\n",
    "    # y=gos.Y(\"gc_content:Q\", domain=(0, 50), flip=False),\n",
    "    tooltip=[\"position:G\", \"gc_content:Q\"],\n",
    ")\n",
    "\n",
    "track_gc_skew = gos.Track(data_gc_skew).mark_line().encode(\n",
    "    # color=gos.value(\"#ff4000\"),\n",
    "    x=gos.X(\"position:G\"),\n",
    "    # y=gos.Y(\"depth:Q\", flip=False),\n",
    "    tooltip=[\"position:G\", \"gc_skew:Q\"]\n",
    ")\n",
    "\n",
    "track_coverage = gos.Track(data_coverage).mark_area().encode(\n",
    "    x=gos.X(\"position:G\", axis=\"bottom\"),\n",
    "    # y=gos.Y(\"depth:Q\", flip=False),\n",
    "    tooltip=[\"position:G\", \"depth:Q\"]\n",
    ")\n",
    "\n",
    "track_gene_posstrand_label = gos.overlay(\n",
    "    track_gene_posstrand, label_gene_posstrand\n",
    ").properties(\n",
    "    width=500, \n",
    "    height=10,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    layout=\"circular\",\n",
    ")\n",
    "\n",
    "track_gene_negstrand_label = gos.overlay(\n",
    "    track_gene_negstrand, label_gene_negstrand\n",
    ").properties(\n",
    "    width=500, \n",
    "    height=10,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    layout=\"circular\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8dc241b",
   "metadata": {},
   "source": [
    "## Circos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a0484b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-ae5790ebafc54d7cb417532e9ddf386c\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@0.17.0?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-ae5790ebafc54d7cb417532e9ddf386c');\n",
       "    let spec = {\"alignment\": \"stack\", \"tracks\": [{\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"rect\", \"color\": {\"domain\": [\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"], \"field\": \"gene_prefix\", \"range\": [\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"], \"type\": \"nominal\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"], \"not\": true}], \"stroke\": {\"value\": \"#000000\"}, \"strokeWidth\": {\"value\": 0.5}, \"title\": \"Positive Strand\", \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"x\": {\"axis\": \"none\", \"domain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15000]}, \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"text\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"], \"not\": true}], \"size\": {\"value\": 15}, \"style\": {\"dy\": 15}, \"text\": {\"field\": \"gene\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"visibility\": [{\"operation\": \"LT\", \"measure\": \"width\", \"threshold\": \"|xe-x|\", \"transitionPadding\": 10, \"target\": \"mark\"}], \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}], \"width\": 500, \"height\": 10, \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}, \"layout\": \"circular\"}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"rect\", \"color\": {\"domain\": [\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"], \"field\": \"gene_prefix\", \"range\": [\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"], \"type\": \"nominal\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"], \"not\": true}], \"stroke\": {\"value\": \"#000000\"}, \"strokeWidth\": {\"value\": 0.5}, \"title\": \"Negative Strand\", \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"x\": {\"axis\": \"none\", \"domain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15000]}, \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"text\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"], \"not\": true}], \"size\": {\"value\": 15}, \"style\": {\"dy\": 15}, \"text\": {\"field\": \"gene\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"visibility\": [{\"operation\": \"LT\", \"measure\": \"width\", \"threshold\": \"|xe-x|\", \"transitionPadding\": 10, \"target\": \"mark\"}], \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}], \"width\": 500, \"height\": 10, \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}, \"layout\": \"circular\"}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/0e7065f-MZ781228_gc_content.csv\"}, \"mark\": \"area\", \"color\": {\"value\": \"steelBlue\"}, \"height\": 10, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"gc_content\", \"type\": \"quantitative\"}], \"width\": 500, \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"gc_content\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/18c411e-MZ781228_gc_skew.csv\"}, \"mark\": \"line\", \"height\": 10, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"gc_skew\", \"type\": \"quantitative\"}], \"width\": 500, \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"domain\": [-1, 1], \"field\": \"gc_skew\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/f9c4d94-MZ781228_sort_out.12F.bam.tsv\", \"separator\": \"\\t\", \"headerNames\": [\"id\", \"position\", \"depth\"]}, \"mark\": \"area\", \"height\": 10, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"depth\", \"type\": \"quantitative\"}], \"width\": 500, \"x\": {\"axis\": \"bottom\", \"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"depth\", \"type\": \"quantitative\"}}], \"title\": \"Basic Marks: Rect\", \"subtitle\": \"Tutorial Examples\", \"layout\": \"circular\", \"xDomain\": {\"interval\": [0, 15235]}, \"spacing\": 0, \"static\": true, \"centerRadius\": 0.3};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'stack',\n",
       "  centerRadius: 0.3,\n",
       "  layout: 'circular',\n",
       "  spacing: 0,\n",
       "  static: True,\n",
       "  subtitle: 'Tutorial Examples',\n",
       "  title: 'Basic Marks: Rect',\n",
       "  tracks: (View({\n",
       "    alignment: 'overlay',\n",
       "    height: 10,\n",
       "    layout: 'circular',\n",
       "    tracks: (Track({\n",
       "      color: Color({\n",
       "        domain: ['nad', 'cox', 'atp', 'trn', 'rrn', 'D-loop'],\n",
       "        field: 'gene_prefix',\n",
       "        range: ['#8c510a', '#d8b365', '#f6e8c3', '#c7eae5', '#5ab4ac', '#01665e'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['-'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rect',\n",
       "      stroke: StrokeValue({\n",
       "        value: '#000000'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 0.5\n",
       "      }),\n",
       "      title: 'Positive Strand',\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'none',\n",
       "        domain: GenomicDomain({\n",
       "          chromosome: 'chr1',\n",
       "          interval: [0, 15000]\n",
       "        }),\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['-'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'text',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      style: Style({\n",
       "        dy: 15\n",
       "      }),\n",
       "      text: Text({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'width',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: '|xe-x|',\n",
       "        transitionPadding: 10\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'top',\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    })),\n",
       "    width: 500,\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: 'chr1',\n",
       "      interval: [0, 15235]\n",
       "    })\n",
       "  }), View({\n",
       "    alignment: 'overlay',\n",
       "    height: 10,\n",
       "    layout: 'circular',\n",
       "    tracks: (Track({\n",
       "      color: Color({\n",
       "        domain: ['nad', 'cox', 'atp', 'trn', 'rrn', 'D-loop'],\n",
       "        field: 'gene_prefix',\n",
       "        range: ['#8c510a', '#d8b365', '#f6e8c3', '#c7eae5', '#5ab4ac', '#01665e'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['+'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rect',\n",
       "      stroke: StrokeValue({\n",
       "        value: '#000000'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 0.5\n",
       "      }),\n",
       "      title: 'Negative Strand',\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'none',\n",
       "        domain: GenomicDomain({\n",
       "          chromosome: 'chr1',\n",
       "          interval: [0, 15000]\n",
       "        }),\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['+'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'text',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      style: Style({\n",
       "        dy: 15\n",
       "      }),\n",
       "      text: Text({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'width',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: '|xe-x|',\n",
       "        transitionPadding: 10\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'top',\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    })),\n",
       "    width: 500,\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: 'chr1',\n",
       "      interval: [0, 15235]\n",
       "    })\n",
       "  }), Track({\n",
       "    color: ColorValue({\n",
       "      value: 'steelBlue'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/0e7065f-MZ781228_gc_content.csv'},\n",
       "    height: 10,\n",
       "    mark: 'area',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'gc_content',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 500,\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'gc_content',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/18c411e-MZ781228_gc_skew.csv'},\n",
       "    height: 10,\n",
       "    mark: 'line',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'gc_skew',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 500,\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      domain: (-1, 1),\n",
       "      field: 'gc_skew',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/f9c4d94-MZ781228_sort_out.12F.bam.tsv', 'separator': '\\t', 'headerNames': ['id', 'position', 'depth']},\n",
       "    height: 10,\n",
       "    mark: 'area',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'depth',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 500,\n",
       "    x: X({\n",
       "      axis: 'bottom',\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'depth',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })),\n",
       "  xDomain: GenomicDomain({\n",
       "    interval: [0, 15235]\n",
       "  })\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "track_gene_posstrand = track_gene_posstrand.properties(width=500, height=10)\n",
    "track_gene_negstrand = track_gene_negstrand.properties(width=500, height=10)\n",
    "track_gc = track_gc.properties(y=gos.Y(\"gc_content:Q\"), width= 500, height=10)\n",
    "track_gc_skew = track_gc_skew.properties(y=gos.Y(\"gc_skew:Q\", domain=(-1, 1)), width= 500, height=10)\n",
    "track_coverage = track_coverage.properties(y=gos.Y(\"depth:Q\"), width= 500, height=10)\n",
    "\n",
    "vis = gos.stack(track_gene_posstrand_label, track_gene_negstrand_label, track_gc, track_gc_skew, track_coverage).properties(\n",
    "    title=\"Basic Marks: Rect\",\n",
    "    subtitle=\"Tutorial Examples\",\n",
    "    layout=\"circular\",\n",
    "    xDomain=gos.GenomicDomain(interval=[0, 15235]),\n",
    "    spacing=0,\n",
    "    static=True,\n",
    "    centerRadius=0.3,\n",
    ")\n",
    "\n",
    "vis.display()\n",
    "vis.save(\"circular_layout.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "497ee420",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-9e24371a4897479da38fb87a00c4f230\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@0.17.0?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-9e24371a4897479da38fb87a00c4f230');\n",
       "    let spec = {\"alignment\": \"stack\", \"tracks\": [{\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"rect\", \"color\": {\"domain\": [\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"], \"field\": \"gene_prefix\", \"range\": [\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"], \"type\": \"nominal\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"], \"not\": true}], \"stroke\": {\"value\": \"#000000\"}, \"strokeWidth\": {\"value\": 0.5}, \"title\": \"Positive Strand\", \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"x\": {\"axis\": \"none\", \"domain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15000]}, \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"text\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"], \"not\": true}], \"size\": {\"value\": 15}, \"style\": {\"dy\": 15}, \"text\": {\"field\": \"gene\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"visibility\": [{\"operation\": \"LT\", \"measure\": \"width\", \"threshold\": \"|xe-x|\", \"transitionPadding\": 10, \"target\": \"mark\"}], \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}], \"width\": 500, \"height\": 10, \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}, \"layout\": \"circular\"}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"rect\", \"color\": {\"domain\": [\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"], \"field\": \"gene_prefix\", \"range\": [\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"], \"type\": \"nominal\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"], \"not\": true}], \"stroke\": {\"value\": \"#000000\"}, \"strokeWidth\": {\"value\": 0.5}, \"title\": \"Negative Strand\", \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"x\": {\"axis\": \"none\", \"domain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15000]}, \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/af9e891-MZ781228.gff3.csv\", \"chromosomeField\": \"seq_id\", \"chromosomePrefix\": \"MZ781228.\", \"genomicFields\": [\"start\", \"end\"]}, \"mark\": \"text\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"gbkey\", \"oneOf\": [\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"], \"not\": true}], \"size\": {\"value\": 15}, \"style\": {\"dy\": 15}, \"text\": {\"field\": \"gene\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"start\", \"type\": \"genomic\"}, {\"field\": \"end\", \"type\": \"genomic\"}, {\"field\": \"strand\", \"type\": \"nominal\"}, {\"field\": \"gene\", \"type\": \"nominal\"}], \"visibility\": [{\"operation\": \"LT\", \"measure\": \"width\", \"threshold\": \"|xe-x|\", \"transitionPadding\": 10, \"target\": \"mark\"}], \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}], \"width\": 500, \"height\": 10, \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}, \"layout\": \"circular\"}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/0e7065f-MZ781228_gc_content.csv\"}, \"mark\": \"area\", \"color\": {\"value\": \"steelBlue\"}, \"height\": 50, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"gc_content\", \"type\": \"quantitative\"}], \"width\": 1000, \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"domain\": [0, 50], \"field\": \"gc_content\", \"flip\": false, \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/18c411e-MZ781228_gc_skew.csv\"}, \"mark\": \"line\", \"height\": 50, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"gc_skew\", \"type\": \"quantitative\"}], \"width\": 1000, \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"domain\": [-0.7, 0.1], \"field\": \"gc_skew\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/f9c4d94-MZ781228_sort_out.12F.bam.tsv\", \"separator\": \"\\t\", \"headerNames\": [\"id\", \"position\", \"depth\"]}, \"mark\": \"area\", \"height\": 50, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"depth\", \"type\": \"quantitative\"}], \"width\": 1000, \"x\": {\"axis\": \"bottom\", \"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"depth\", \"flip\": false, \"type\": \"quantitative\"}}], \"title\": \"Basic Marks: Rect\", \"subtitle\": \"Tutorial Examples\", \"layout\": \"linear\", \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}, \"spacing\": 0, \"static\": false};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'stack',\n",
       "  layout: 'linear',\n",
       "  spacing: 0,\n",
       "  static: False,\n",
       "  subtitle: 'Tutorial Examples',\n",
       "  title: 'Basic Marks: Rect',\n",
       "  tracks: (View({\n",
       "    alignment: 'overlay',\n",
       "    height: 10,\n",
       "    layout: 'circular',\n",
       "    tracks: (Track({\n",
       "      color: Color({\n",
       "        domain: ['nad', 'cox', 'atp', 'trn', 'rrn', 'D-loop'],\n",
       "        field: 'gene_prefix',\n",
       "        range: ['#8c510a', '#d8b365', '#f6e8c3', '#c7eae5', '#5ab4ac', '#01665e'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['-'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rect',\n",
       "      stroke: StrokeValue({\n",
       "        value: '#000000'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 0.5\n",
       "      }),\n",
       "      title: 'Positive Strand',\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'none',\n",
       "        domain: GenomicDomain({\n",
       "          chromosome: 'chr1',\n",
       "          interval: [0, 15000]\n",
       "        }),\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['-'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'text',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      style: Style({\n",
       "        dy: 15\n",
       "      }),\n",
       "      text: Text({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'width',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: '|xe-x|',\n",
       "        transitionPadding: 10\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'top',\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    })),\n",
       "    width: 500,\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: 'chr1',\n",
       "      interval: [0, 15235]\n",
       "    })\n",
       "  }), View({\n",
       "    alignment: 'overlay',\n",
       "    height: 10,\n",
       "    layout: 'circular',\n",
       "    tracks: (Track({\n",
       "      color: Color({\n",
       "        domain: ['nad', 'cox', 'atp', 'trn', 'rrn', 'D-loop'],\n",
       "        field: 'gene_prefix',\n",
       "        range: ['#8c510a', '#d8b365', '#f6e8c3', '#c7eae5', '#5ab4ac', '#01665e'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['+'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rect',\n",
       "      stroke: StrokeValue({\n",
       "        value: '#000000'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 0.5\n",
       "      }),\n",
       "      title: 'Negative Strand',\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'none',\n",
       "        domain: GenomicDomain({\n",
       "          chromosome: 'chr1',\n",
       "          interval: [0, 15000]\n",
       "        }),\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/af9e891-MZ781228.gff3.csv', 'chromosomeField': 'seq_id', 'chromosomePrefix': 'MZ781228.', 'genomicFields': ['start', 'end']},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'gbkey',\n",
       "        oneOf: ['tRNA', 'rRNA', 'CDS', 'D-loop'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        not: True,\n",
       "        oneOf: ['+'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'text',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      style: Style({\n",
       "        dy: 15\n",
       "      }),\n",
       "      text: Text({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      tooltip: [Tooltip({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }), Tooltip({\n",
       "        field: 'strand',\n",
       "        type: 'nominal'\n",
       "      }), Tooltip({\n",
       "        field: 'gene',\n",
       "        type: 'nominal'\n",
       "      })],\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'width',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: '|xe-x|',\n",
       "        transitionPadding: 10\n",
       "      })],\n",
       "      x: X({\n",
       "        axis: 'top',\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    })),\n",
       "    width: 500,\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: 'chr1',\n",
       "      interval: [0, 15235]\n",
       "    })\n",
       "  }), Track({\n",
       "    color: ColorValue({\n",
       "      value: 'steelBlue'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/0e7065f-MZ781228_gc_content.csv'},\n",
       "    height: 50,\n",
       "    mark: 'area',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'gc_content',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 1000,\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      domain: (0, 50),\n",
       "      field: 'gc_content',\n",
       "      flip: False,\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/18c411e-MZ781228_gc_skew.csv'},\n",
       "    height: 50,\n",
       "    mark: 'line',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'gc_skew',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 1000,\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      domain: (-0.7, 0.1),\n",
       "      field: 'gc_skew',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/f9c4d94-MZ781228_sort_out.12F.bam.tsv', 'separator': '\\t', 'headerNames': ['id', 'position', 'depth']},\n",
       "    height: 50,\n",
       "    mark: 'area',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'depth',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 1000,\n",
       "    x: X({\n",
       "      axis: 'bottom',\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'depth',\n",
       "      flip: False,\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })),\n",
       "  xDomain: GenomicDomain({\n",
       "    chromosome: 'chr1',\n",
       "    interval: [0, 15235]\n",
       "  })\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "track_gene_posstrand = track_gene_posstrand.properties(width=1000, height=25)\n",
    "track_gene_negstrand = track_gene_negstrand.properties(width=1000, height=25)\n",
    "track_gc = track_gc.properties(y=gos.Y(\"gc_content:Q\", domain=(0, 50), flip=False), width=1000, height=50)\n",
    "track_gc_skew = track_gc_skew.properties(y=gos.Y(\"gc_skew:Q\", domain=(-0.7, 0.1)), width= 1000, height=50)\n",
    "track_coverage = track_coverage.properties(y=gos.Y(\"depth:Q\", flip=False), width=1000, height=50)\n",
    "\n",
    "track_base_2 = gos.Track(data_base).mark_bar().encode(\n",
    "    x=gos.X(\"Start:G\"),\n",
    "    xe=gos.X(\"End:G\"),\n",
    "    y=gos.Y(\"Count:Q\", axis=\"none\"),\n",
    "    color=gos.Color(\"Base:N\", domain=[\"A\", \"T\", \"G\", \"C\"], legend=True)\n",
    ")\n",
    "\n",
    "track_base_text_2 = gos.Track(data_base).mark_text().encode(\n",
    "    x=gos.X(\"Start:G\"),\n",
    "    xe=gos.X(\"End:G\"),\n",
    "    size=gos.value(24),\n",
    "    color=gos.value(\"white\"),\n",
    "    text=\"Base:N\",\n",
    ").visibility_lt(\n",
    "    measure=\"width\", threshold=\"|xe-x|\", transitionPadding=30, target=\"mark\"\n",
    ").visibility_lt(\n",
    "    measure=\"zoomLevel\", threshold=10, target=\"track\"\n",
    ").transform_filter_not(\"Count\", oneOf=[0])\n",
    "\n",
    "track_gene_base_label = gos.overlay(\n",
    "    track_base_2, track_base_text_2\n",
    ").properties(\n",
    "    width=1000, \n",
    "    height=25,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    ")\n",
    "\n",
    "vis = gos.stack(track_gene_posstrand_label, track_gene_negstrand_label, track_gc, track_gc_skew, track_coverage).properties(\n",
    "    title=\"Basic Marks: Rect\",\n",
    "    subtitle=\"Tutorial Examples\",\n",
    "    layout=\"linear\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    spacing=0,\n",
    "    static=False,\n",
    ")\n",
    "\n",
    "vis.display()\n",
    "vis.save(\"linear_layout.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4884f850",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-0424d933986c4488ba6d156c0c78aca8\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@0.17.0?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-0424d933986c4488ba6d156c0c78aca8');\n",
       "    let spec = {\"alignment\": \"stack\", \"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/0e7065f-MZ781228_gc_content.csv\"}, \"mark\": \"area\", \"color\": {\"value\": \"steelBlue\"}, \"height\": 50, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"gc_content\", \"type\": \"quantitative\"}], \"width\": 1000, \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"domain\": [0, 50], \"field\": \"gc_content\", \"flip\": false, \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/18c411e-MZ781228_gc_skew.csv\"}, \"mark\": \"line\", \"height\": 50, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"gc_skew\", \"type\": \"quantitative\"}], \"width\": 1000, \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"domain\": [-0.7, 0.1], \"field\": \"gc_skew\", \"type\": \"quantitative\"}}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/2b73a8b-MZ781228.fasta.csv\"}, \"mark\": \"bar\", \"color\": {\"domain\": [\"A\", \"T\", \"G\", \"C\"], \"field\": \"Base\", \"legend\": true, \"type\": \"nominal\"}, \"x\": {\"field\": \"Start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"End\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"Count\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/2b73a8b-MZ781228.fasta.csv\"}, \"mark\": \"text\", \"color\": {\"value\": \"white\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"Count\", \"oneOf\": [0], \"not\": true}], \"size\": {\"value\": 24}, \"text\": {\"field\": \"Base\", \"type\": \"nominal\"}, \"visibility\": [{\"operation\": \"LT\", \"measure\": \"width\", \"threshold\": \"|xe-x|\", \"transitionPadding\": 30, \"target\": \"mark\"}, {\"operation\": \"LT\", \"measure\": \"zoomLevel\", \"threshold\": 10, \"target\": \"track\"}], \"x\": {\"field\": \"Start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"End\", \"type\": \"genomic\"}}], \"width\": 1000, \"height\": 25, \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}}, {\"data\": {\"type\": \"csv\", \"url\": \"http://localhost:32771/resources/f9c4d94-MZ781228_sort_out.12F.bam.tsv\", \"separator\": \"\\t\", \"headerNames\": [\"id\", \"position\", \"depth\"]}, \"mark\": \"area\", \"height\": 50, \"tooltip\": [{\"field\": \"position\", \"type\": \"genomic\"}, {\"field\": \"depth\", \"type\": \"quantitative\"}], \"width\": 1000, \"x\": {\"axis\": \"bottom\", \"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"depth\", \"flip\": false, \"type\": \"quantitative\"}}], \"title\": \"Genomic Properties + Sequence Alignment\", \"subtitle\": \"Hybrid Layout\", \"layout\": \"linear\", \"xDomain\": {\"chromosome\": \"chr1\", \"interval\": [0, 15235]}, \"spacing\": 0, \"static\": false};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'stack',\n",
       "  layout: 'linear',\n",
       "  spacing: 0,\n",
       "  static: False,\n",
       "  subtitle: 'Hybrid Layout',\n",
       "  title: 'Genomic Properties + Sequence Alignment',\n",
       "  tracks: (Track({\n",
       "    color: ColorValue({\n",
       "      value: 'steelBlue'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/0e7065f-MZ781228_gc_content.csv'},\n",
       "    height: 50,\n",
       "    mark: 'area',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'gc_content',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 1000,\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      domain: (0, 50),\n",
       "      field: 'gc_content',\n",
       "      flip: False,\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/18c411e-MZ781228_gc_skew.csv'},\n",
       "    height: 50,\n",
       "    mark: 'line',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'gc_skew',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 1000,\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      domain: (-0.7, 0.1),\n",
       "      field: 'gc_skew',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), View({\n",
       "    alignment: 'overlay',\n",
       "    height: 25,\n",
       "    tracks: (Track({\n",
       "      color: Color({\n",
       "        domain: ['A', 'T', 'G', 'C'],\n",
       "        field: 'Base',\n",
       "        legend: True,\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/2b73a8b-MZ781228.fasta.csv'},\n",
       "      mark: 'bar',\n",
       "      x: X({\n",
       "        field: 'Start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: X({\n",
       "        field: 'End',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        axis: 'none',\n",
       "        field: 'Count',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    }), Track({\n",
       "      color: ColorValue({\n",
       "        value: 'white'\n",
       "      }),\n",
       "      data: {'type': 'csv', 'url': 'http://localhost:32771/resources/2b73a8b-MZ781228.fasta.csv'},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'Count',\n",
       "        not: True,\n",
       "        oneOf: [0],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'text',\n",
       "      size: SizeValue({\n",
       "        value: 24\n",
       "      }),\n",
       "      text: Text({\n",
       "        field: 'Base',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'width',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: '|xe-x|',\n",
       "        transitionPadding: 30\n",
       "      }), VisibilityCondition({\n",
       "        measure: 'zoomLevel',\n",
       "        operation: 'LT',\n",
       "        target: 'track',\n",
       "        threshold: 10\n",
       "      })],\n",
       "      x: X({\n",
       "        field: 'Start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: X({\n",
       "        field: 'End',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    })),\n",
       "    width: 1000,\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: 'chr1',\n",
       "      interval: [0, 15235]\n",
       "    })\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'http://localhost:32771/resources/f9c4d94-MZ781228_sort_out.12F.bam.tsv', 'separator': '\\t', 'headerNames': ['id', 'position', 'depth']},\n",
       "    height: 50,\n",
       "    mark: 'area',\n",
       "    tooltip: [Tooltip({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }), Tooltip({\n",
       "      field: 'depth',\n",
       "      type: 'quantitative'\n",
       "    })],\n",
       "    width: 1000,\n",
       "    x: X({\n",
       "      axis: 'bottom',\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'depth',\n",
       "      flip: False,\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })),\n",
       "  xDomain: GenomicDomain({\n",
       "    chromosome: 'chr1',\n",
       "    interval: [0, 15235]\n",
       "  })\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "track_gene_posstrand = track_gene_posstrand.properties(width=1000, height=25)\n",
    "track_gene_negstrand = track_gene_negstrand.properties(width=1000, height=25)\n",
    "track_gc = track_gc.properties(y=gos.Y(\"gc_content:Q\", domain=(0, 50), flip=False), width=1000, height=50)\n",
    "track_gc_skew = track_gc_skew.properties(y=gos.Y(\"gc_skew:Q\", domain=(-0.7, 0.1)), width= 1000, height=50)\n",
    "track_coverage = track_coverage.properties(y=gos.Y(\"depth:Q\", flip=False), width=1000, height=50)\n",
    "\n",
    "track_base_2 = gos.Track(data_base).mark_bar().encode(\n",
    "    x=gos.X(\"Start:G\"),\n",
    "    xe=gos.X(\"End:G\"),\n",
    "    y=gos.Y(\"Count:Q\", axis=\"none\"),\n",
    "    color=gos.Color(\"Base:N\", domain=[\"A\", \"T\", \"G\", \"C\"], legend=True)\n",
    ")\n",
    "\n",
    "track_base_text_2 = gos.Track(data_base).mark_text().encode(\n",
    "    x=gos.X(\"Start:G\"),\n",
    "    xe=gos.X(\"End:G\"),\n",
    "    size=gos.value(24),\n",
    "    color=gos.value(\"white\"),\n",
    "    text=\"Base:N\",\n",
    ").visibility_lt(\n",
    "    measure=\"width\", threshold=\"|xe-x|\", transitionPadding=30, target=\"mark\"\n",
    ").visibility_lt(\n",
    "    measure=\"zoomLevel\", threshold=10, target=\"track\"\n",
    ").transform_filter_not(\"Count\", oneOf=[0])\n",
    "\n",
    "track_gene_base_label = gos.overlay(\n",
    "    track_base_2, track_base_text_2\n",
    ").properties(\n",
    "    width=1000, \n",
    "    height=25,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    ")\n",
    "\n",
    "vis = gos.stack(track_gc, track_gc_skew, track_gene_base_label, track_coverage).properties(\n",
    "    title=\"Genomic Properties + Sequence Alignment\",\n",
    "    subtitle=\"Hybrid Layout\",\n",
    "    layout=\"linear\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    spacing=0,\n",
    "    static=False,\n",
    ")\n",
    "\n",
    "vis.display()\n",
    "vis.save(\"linear_layout_2.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c739e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "track = gos.Track(data).mark_rect().encode(\n",
    "    color=gos.Color(\"gene_prefix:N\",\n",
    "            domain=[\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"],\n",
    "            range=[\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"-\"]\n",
    ").properties(width=500, height=10)\n",
    "\n",
    "track_reverse = gos.Track(data).mark_rect().encode(\n",
    "    color=gos.Color(\"gbkey:N\",\n",
    "            domain=[\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"],\n",
    "            range=[\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"+\"]\n",
    ").properties(width=500, height=10)\n",
    "\n",
    "track_gc = gos.Track(data_gc).mark_bar().encode(\n",
    "    x=gos.X(\"position:G\"),\n",
    "    # xe=gos.X(\"position:G\", offset=10),\n",
    "    y=gos.Y(\"gc_content:Q\", domain=(0, 50), flip=True),\n",
    "    color=gos.Color(\n",
    "        field=\"gc_content\",\n",
    "        domain=(10, 20, 30),\n",
    "        range=[\"#1E88E5\", \"#7CB342\", \"#D81B60\"]\n",
    "    ),\n",
    "    tooltip=[\"position:G\", \"gc_content:Q\"],\n",
    ").properties(width=500, height=10)\n",
    "\n",
    "track_coverage = gos.Track(data_coverage).mark_line().encode(\n",
    "    x=gos.X(\"position:G\", axis=\"bottom\"),\n",
    "    y=gos.Y(\"depth:Q\", flip=True),\n",
    "    tooltip=[\"position:G\", \"depth:Q\"]\n",
    ").properties(width= 500, height=10)\n",
    "\n",
    "vis = gos.stack(track, track_reverse, track_gc, track_coverage).properties(\n",
    "    title=\"Basic Marks: Rect\",\n",
    "    subtitle=\"Tutorial Examples\",\n",
    "    layout=\"circular\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    spacing=0,\n",
    "    static=True,\n",
    "    centerRadius=0.3,\n",
    ")\n",
    "\n",
    "vis.display()\n",
    "vis.save(\"circular_layout.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9de965c",
   "metadata": {},
   "source": [
    "### Circos Test 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf9ebc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "base = gos.Track(data).encode(\n",
    "    # rows=gos.Row(\"strand:N\", domain=[\"+\", \"-\"]),\n",
    "    tooltip=[\"start:G\", \"end:G\", \"strand:N\", \"gene:N\"]\n",
    ").properties(\n",
    "    title=\"Genes | MZ781228\"\n",
    ")\n",
    "\n",
    "track = base.mark_rect().encode(\n",
    "    color=gos.Color(\"gene_prefix:N\",\n",
    "            domain=[\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"],\n",
    "            range=[\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ")\n",
    "\n",
    "track_reverse = gos.Track(data).mark_rect().encode(\n",
    "    color=gos.Color(\"gbkey:N\",\n",
    "            domain=[\"nad\", \"cox\", \"atp\", \"trn\", \"rrn\", \"D-loop\"],\n",
    "            range=[\"#8c510a\", \"#d8b365\", \"#f6e8c3\", \"#c7eae5\", \"#5ab4ac\", \"#01665e\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"+\"]\n",
    ").properties(width=500, height=10)\n",
    "\n",
    "gene_label = base.mark_text(dy=15).encode(\n",
    "    x=gos.X(\"start:G\", axis=\"top\"),\n",
    "    xe=\"end:G\",\n",
    "    text=\"gene:N\",\n",
    "    size=gos.value(15)\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").visibility_lt(\n",
    "    measure=\"width\",\n",
    "    threshold=\"|xe-x|\",\n",
    "    transitionPadding=10,\n",
    "    target=\"mark\",\n",
    ")\n",
    "\n",
    "track_gc = gos.Track(data_gc).mark_bar().encode(\n",
    "    x=gos.X(\"position:G\"),\n",
    "    # xe=gos.X(\"position:G\", offset=10),\n",
    "    y=gos.Y(\"gc_content:Q\", domain=(0, 50), flip=True),\n",
    "    color=gos.Color(\n",
    "        field=\"gc_content\",\n",
    "        domain=(10, 20, 30),\n",
    "        range=[\"#1E88E5\", \"#7CB342\", \"#D81B60\"]\n",
    "    ),\n",
    "    tooltip=[\"position:G\", \"gc_content:Q\"],\n",
    ").properties(width=500, height=10)\n",
    "\n",
    "track_coverage = gos.Track(data_coverage).mark_line().encode(\n",
    "    x=gos.X(\"position:G\", axis=\"bottom\"),\n",
    "    y=gos.Y(\"depth:Q\", flip=True),\n",
    "    tooltip=[\"position:G\", \"depth:Q\"]\n",
    ").properties(width= 500, height=10)\n",
    "\n",
    "track_gene = gos.overlay(\n",
    "    track, gene_label\n",
    ").properties(\n",
    "    width=500, \n",
    "    height=10,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    layout=\"circular\",\n",
    ")\n",
    "\n",
    "vis = gos.stack(track_gene, track_gc, track_coverage).properties(\n",
    "    title=\"Basic Marks: Rect\",\n",
    "    subtitle=\"Tutorial Examples\",\n",
    "    layout=\"circular\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    spacing=0,\n",
    "    static=True,\n",
    "    centerRadius=0.3,\n",
    ")\n",
    "\n",
    "vis.display()\n",
    "# vis.save(\"circular_layout.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39ce4c3b",
   "metadata": {},
   "source": [
    "## Linear Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7648ec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "track_2 = gos.Track(data).mark_rect().encode(\n",
    "    color=gos.Color(\"gbkey:N\",\n",
    "            domain=[\"CDS\", \"tRNA\", \"rRNA\", \"D-loop\"],\n",
    "            range=[\"#ffff00\", \"#000080\", \"#ff4500\", \"#000000\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"-\"]\n",
    ").properties(width=1000, height=25)\n",
    "\n",
    "track_reverse_2 = gos.Track(data).mark_rect().encode(\n",
    "    color=gos.Color(\"gbkey:N\",\n",
    "            domain=[\"CDS\", \"tRNA\", \"rRNA\", \"D-loop\"],\n",
    "            range=[\"#ffff00\", \"#000080\", \"#ff4500\", \"#000000\"],\n",
    "        ),\n",
    "    x=gos.X(\"start:G\", domain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15000]), axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    stroke=gos.value(\"#000000\"),\n",
    "    strokeWidth=gos.value(0.5),\n",
    ").transform_filter(\n",
    "    \"gbkey\", \n",
    "    oneOf=[\"tRNA\", \"rRNA\", \"CDS\", \"D-loop\"]\n",
    ").transform_filter_not(\n",
    "    \"strand\", \n",
    "    oneOf=[\"+\"]\n",
    ").properties(width=1000, height=25)\n",
    "\n",
    "track_base_2 = gos.Track(data_base).mark_bar().encode(\n",
    "    x=gos.X(\"Start:G\"),\n",
    "    xe=gos.X(\"End:G\"),\n",
    "    y=gos.Y(\"Count:Q\", axis=\"none\"),\n",
    "    color=gos.Color(\"Base:N\", domain=[\"A\", \"T\", \"G\", \"C\"], legend=True)\n",
    ").properties(width=1000, height=50)\n",
    "\n",
    "track_base_text_2 = gos.Track(data_base).mark_text().encode(\n",
    "    x=gos.X(\"Start:G\"),\n",
    "    xe=gos.X(\"End:G\"),\n",
    "    size=gos.value(24),\n",
    "    color=gos.value(\"white\"),\n",
    "    text=\"Base:N\",\n",
    ").visibility_lt(\n",
    "    measure=\"width\", threshold=\"|xe-x|\", transitionPadding=30, target=\"mark\"\n",
    ").visibility_lt(\n",
    "    measure=\"zoomLevel\", threshold=10, target=\"track\"\n",
    ").transform_filter_not(\"Count\", oneOf=[0])\n",
    "\n",
    "track_gc_2 = gos.Track(data_gc).mark_bar().encode(\n",
    "    x=gos.X(\"position:G\"),\n",
    "    # xe=gos.X(\"position:G\", offset=10),\n",
    "    y=gos.Y(\"gc_content:Q\", domain=(0, 50)),\n",
    "    color=gos.Color(\n",
    "        field=\"gc_content\",\n",
    "        domain=(10, 20, 30),\n",
    "        range=[\"#1E88E5\", \"#7CB342\", \"#D81B60\"]\n",
    "    ),\n",
    "    tooltip=[\"position:G\", \"gc_content:Q\"],\n",
    ").properties(width=1000, height=50)\n",
    "\n",
    "track_coverage_2 = gos.Track(data_coverage).mark_line().encode(\n",
    "    x=gos.X(\"position:G\", axis=\"bottom\"),\n",
    "    y=gos.Y(\"depth:Q\"),\n",
    "    tooltip=[\"position:G\", \"depth:Q\"]\n",
    ").properties(width= 1000, height=50)\n",
    "\n",
    "vis_base_text = gos.overlay(track_base_2, track_base_text_2).properties(\n",
    "    title=\"Multi-Scale Sequence Plot\",\n",
    "    xDomain=gos.GenomicDomain(interval=[0, 15235]),\n",
    "    width=725,\n",
    "    height=100\n",
    ")\n",
    "\n",
    "vis_2 = gos.stack(track_2, track_reverse_2, vis_base_text, track_gc_2, track_coverage_2).properties(\n",
    "    title=\"Basic Marks: Rect\",\n",
    "    subtitle=\"Tutorial Examples\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\", interval=[0, 15235]),\n",
    "    spacing=0,\n",
    "    # static=True,\n",
    "    centerRadius=0.3,\n",
    ")\n",
    "\n",
    "vis_2.display()\n",
    "vis_2.save(\"linear_layout.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12bb2d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "vis_base_text.display()\n",
    "vis_base_text.save(\"linear_layout_base.html\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "seqaln-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
